File: app/src/main/java/com/djacoronel/sleeklauncher/home/MainActivity.java
package com.djacoronel.sleeklauncher.home;

import android.app.Activity;
import android.app.ActivityOptions;
import android.app.AlertDialog;
import android.app.WallpaperManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.View;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.Toast;

import com.djacoronel.sleeklauncher.R;
import com.djacoronel.sleeklauncher.data.model.AppDetail;
import com.djacoronel.sleeklauncher.data.model.IconPrefs;
import com.djacoronel.sleeklauncher.data.room.IconPrefsDao;
import com.djacoronel.sleeklauncher.iconutils.IconPackManager;
import com.djacoronel.sleeklauncher.iconutils.IconsActivity;
import com.djacoronel.sleeklauncher.settings.SettingsActivity;
import com.djacoronel.sleeklauncher.settings.BlurBuilder;

import dagger.android.AndroidInjection;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import javax.inject.Inject;

public class MainActivity extends Activity {

  private List<AppDetail> apps;
  GridAdapter adapter;

  @Inject
  IconPrefsDao iconPrefsDao;
  @Inject
  SharedPreferences preferences;

  int selectedAppPosition;
  ImageView image;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    AndroidInjection.inject(this);
    setContentView(R.layout.activity_main);

    loadBackground();
    setupBackgroundRefreshing();
    loadApps();
    loadAppGrid();
    setupGridRefreshing();

    // This makes status bar and navigation bar transparent
    getWindow().setFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
        WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
  }

  void loadBackground() {
    String bgPref = preferences.getString("bgPref", "0,0,0,0,0");
    String argbBlur[] = bgPref.split(",");
    int argb[] = {Integer.parseInt(argbBlur[0]),
        Integer.parseInt(argbBlur[1]),
        Integer.parseInt(argbBlur[2]),
        Integer.parseInt(argbBlur[3])};
    int blur = Integer.parseInt(argbBlur[4]);

    if (blur == 0) {
      loadBackgroundWithColorFilter(argb);
    } else {
      loadBackgroundWithBlur(argb, blur);
    }
  }

  void loadBackgroundWithColorFilter(int argb[]) {
    ImageView mainBg = findViewById(R.id.mainBackground);
    mainBg.setVisibility(View.GONE);
    getWindow().getDecorView().setBackgroundColor(Color.argb(argb[0], argb[1], argb[2], argb[3]));
  }

  void loadBackgroundWithBlur(int argb[], int blur) {
    ImageView mainBg = findViewById(R.id.mainBackground);
    Drawable wallpaperDrawable = WallpaperManager.getInstance(this).getDrawable();
    Bitmap blurredBitmap = cropWallpaper(((BitmapDrawable) wallpaperDrawable).getBitmap());

    int numberOfFullBlur = blur / 25;
    int remainingBlur = blur % 25;

    for (int i = 0; i < numberOfFullBlur; i++) {
      blurredBitmap = new BlurBuilder().blur(this, blurredBitmap, 25);
    }

    if (remainingBlur != 0) {
      blurredBitmap = new BlurBuilder().blur(this, blurredBitmap, remainingBlur);
    }

    mainBg.setVisibility(View.VISIBLE);
    mainBg.setColorFilter(Color.argb(argb[0], argb[1], argb[2], argb[3]));
    mainBg.setImageDrawable(new BitmapDrawable(getResources(), blurredBitmap));
    mainBg.setScaleType(ImageView.ScaleType.FIT_CENTER);
  }

  Bitmap cropWallpaper(Bitmap wallpaper) {
    DisplayMetrics displaymetrics = new DisplayMetrics();
    getWindowManager().getDefaultDisplay().getRealMetrics(displaymetrics);
    int height = displaymetrics.heightPixels;
    int width = displaymetrics.widthPixels;

    return Bitmap.createBitmap(wallpaper, 0, 0, width, height);
  }

  void setupBackgroundRefreshing() {
    IntentFilter intentFilter = new IntentFilter();
    intentFilter.addAction(Intent.ACTION_WALLPAPER_CHANGED);
    BroadcastReceiver br = new BroadcastReceiver() {
      @Override
      public void onReceive(Context context, Intent intent) {
        loadBackground();
      }
    };
    this.registerReceiver(br, intentFilter);
  }

  void loadApps() {
    PackageManager manager = this.getPackageManager();

    apps = new ArrayList<>();

    boolean showHidden = preferences.getBoolean("showHidden", false);

    List<ResolveInfo> availableActivities = getInstalledAppsInfo();

    for (ResolveInfo ri : availableActivities) {
      String name = ri.activityInfo.packageName;
      String label = (String) ri.loadLabel(manager);
      Drawable icon = ri.activityInfo.loadIcon(manager);

      IconPrefs iconPrefs = iconPrefsDao.getIconPrefs(name);

      icon = getAppIcon(name);

      if (iconPrefs != null) {
        if (!iconPrefs.getLabel().equals(IconPrefs.NO_CUSTOM_LABEL)) {
          label = iconPrefs.getLabel();
        }
        if (!iconPrefs.isHidden() || showHidden) {
          AppDetail app = new AppDetail(label, name, icon);
          apps.add(app);
        }
      } else {
        AppDetail app = new AppDetail(label, name, icon);
        apps.add(app);
      }
    }

    Collections.sort(apps, new Comparator<AppDetail>() {
      @Override
      public int compare(AppDetail a1, AppDetail a2) {
        return a1.getLabel().compareToIgnoreCase(a2.getLabel());
      }
    });
  }

  List<ResolveInfo> getInstalledAppsInfo() {
    Intent intent = new Intent(Intent.ACTION_MAIN, null);
    intent.addCategory(Intent.CATEGORY_LAUNCHER);
    return getPackageManager().queryIntentActivities(intent, 0);
  }

  Drawable getAppIcon(String appName) {
    IconPackManager icManager = new IconPackManager(this);

    String selectedIconPack = preferences.getString("iconPack", "");
    HashMap<String, String> icPackComponents = icManager.load(selectedIconPack);

    IconPrefs iconPrefs = iconPrefsDao.getIconPrefs(appName);

    // get custom icon
    if (iconPrefs != null && !iconPrefs.getIconName().equals(IconPrefs.NO_CUSTOM_ICON)) {
      String iconInfo = iconPrefs.getIconName();
      String split[] = iconInfo.split("/");
      String iconDrawable = split[0];
      String iconPackage = split[1];
      return icManager.loadDrawable(iconDrawable, iconPackage);
    }
    // get themed icon if available
    else if (!selectedIconPack.equals("") && icPackComponents.get(appName) != null
        && icManager.loadDrawable(icPackComponents.get(appName), selectedIconPack)
        != null) {
      return icManager.loadDrawable(icPackComponents.get(appName), selectedIconPack);
    } else {
      try {
        return getPackageManager().getApplicationIcon(appName);
      } catch (PackageManager.NameNotFoundException e) {
        e.printStackTrace();
      }
      return null;
    }
  }

  public void loadAppGrid() {
    RecyclerView grid = findViewById(R.id.app_grid);
    adapter = new GridAdapter(apps, this);
    grid.setLayoutManager(new GridLayoutManager(this, 4));
    grid.setAdapter(adapter);
  }

  void setupGridRefreshing() {
    IntentFilter intentFilter = new IntentFilter();
    intentFilter.addAction(Intent.ACTION_PACKAGE_REMOVED);
    intentFilter.addAction(Intent.ACTION_PACKAGE_ADDED);
    intentFilter.addDataScheme("package");
    BroadcastReceiver br = new BroadcastReceiver() {
      @Override
      public void onReceive(Context context, Intent intent) {
        loadApps();
        loadAppGrid();
      }
    };
    this.registerReceiver(br, intentFilter);
  }

  public void launchApp(int position, View v) {
    PackageManager manager = getPackageManager();
    Intent intent = manager.getLaunchIntentForPackage(apps.get(position).getName());
    Bundle optsBundle = getMarshmallowOpeningAnimationOpts(v).toBundle();

    if (intent != null) {
      startActivity(intent, optsBundle);
    } else {
      Toast.makeText(MainActivity.this, "App cannot be found", Toast.LENGTH_SHORT).show();
      apps.remove(position);
      loadAppGrid();
    }
  }

  ActivityOptions getMarshmallowOpeningAnimationOpts(View v) {
    if (Build.VERSION.SDK_INT >= 23) {
      int left = 0, top = 0;
      int width = v.getMeasuredWidth(), height = v.getMeasuredHeight();
      return ActivityOptions.makeClipRevealAnimation(v, left, top, width, height);
    } else {
      // Below L, we use a scale up animation
      return ActivityOptions
          .makeScaleUpAnimation(v, 0, 0, v.getMeasuredWidth(), v.getMeasuredHeight());
    }
  }


  public void openSettings() {
    Intent I = new Intent(this, SettingsActivity.class);
    startActivityForResult(I, 1);
  }

  public void iconLongClick(final int position) {
    selectedAppPosition = position;
    AlertDialog.Builder longClickDialog = buildLongClickDialog();
    longClickDialog = setButtonNameAndActions(longClickDialog);
    longClickDialog = setDialogIcon(longClickDialog);
    longClickDialog.create().show();
  }

  AlertDialog.Builder buildLongClickDialog() {
    return new AlertDialog.Builder(this)
        .setTitle("Options")
        .setMessage("Touch the icon to apply theme.")
        .setCancelable(true)
        .setNegativeButton("Uninstall", new DialogInterface.OnClickListener() {
          @Override
          public void onClick(DialogInterface dialog, int which) {
            Uri packageUri = Uri.parse("package:" + apps.get(selectedAppPosition).getName());
            Intent uninstallIntent = new Intent(Intent.ACTION_DELETE, packageUri);
            startActivity(uninstallIntent);
          }
        });
  }

  AlertDialog.Builder setButtonNameAndActions(AlertDialog.Builder dBuilder) {
    final boolean showHidden = preferences.getBoolean("showHidden", false);
    final String appName = apps.get(selectedAppPosition).getName();
    IconPrefs iconPrefs = iconPrefsDao.getIconPrefs(appName);

    if (iconPrefs == null) {
      iconPrefs = new IconPrefs(appName);
      iconPrefsDao.addIconPrefs(iconPrefs);
    }

    if (iconPrefs.isHidden()) {
      dBuilder.setPositiveButton("Unhide", new DialogInterface.OnClickListener() {
        @Override
        public void onClick(DialogInterface dialog, int which) {
          IconPrefs iconPrefs = iconPrefsDao.getIconPrefs(appName);

          if (iconPrefs != null) {
            iconPrefs.setHidden(false);
            iconPrefsDao.updateIconPrefs(iconPrefs);

            Toast.makeText(MainActivity.this, "App removed from hidden!",
                Toast.LENGTH_LONG).show();
          } else {
            Toast.makeText(MainActivity.this, "App icon preference not found.",
                Toast.LENGTH_LONG).show();
          }
        }
      });
    } else {
      dBuilder.setPositiveButton("Hide", new DialogInterface.OnClickListener() {
        @Override
        public void onClick(DialogInterface dialog, int which) {
          IconPrefs iconPrefs = iconPrefsDao.getIconPrefs(appName);

          if (iconPrefs != null) {
            iconPrefs.setHidden(true);
            iconPrefsDao.updateIconPrefs(iconPrefs);
          }

          if (!showHidden) {
            apps.remove(selectedAppPosition);
            adapter.notifyItemRemoved(selectedAppPosition);
          } else {
            Toast
                .makeText(MainActivity.this, "App marked as hidden, unshow marked apps in settings",
                    Toast.LENGTH_LONG).show();
          }
        }
      });
    }
    return dBuilder;
  }

  AlertDialog.Builder setDialogIcon(AlertDialog.Builder dBuilder) {
    image = new ImageView(this);
    Bitmap bitmap = ((BitmapDrawable) apps.get(selectedAppPosition).getIcon()).getBitmap();
    Drawable icon = new BitmapDrawable(getResources(),
        Bitmap.createScaledBitmap(bitmap, 200, 200, true));
    image.setImageDrawable(icon);
    image.setOnClickListener(
        new View.OnClickListener() {
          @Override
          public void onClick(View v) {
            showIconPackList();
          }
        }
    );
    dBuilder.setView(image);
    return dBuilder;
  }

  public void showIconPackList() {
    IconPackManager icManager = new IconPackManager(this);
    final HashMap<String, String> iconPacks = icManager.getAvailableIconPacks();
    final String[] iconPackNames = iconPacks.keySet().toArray(new String[0]);

    AlertDialog icPackListDialog = buildIconPackList(iconPackNames);
    icPackListDialog.show();
  }

  AlertDialog buildIconPackList(final String[] iconPackNames) {
    return new AlertDialog.Builder(this)
        .setTitle("Icon Packs")
        .setItems(iconPackNames, new DialogInterface.OnClickListener() {
          @Override
          public void onClick(DialogInterface dialog, int which) {
            if (iconPackNames[which].equals("Default")) {
              setDefaultIcon();
            } else {
              String selectedIconPack = iconPackNames[which];
              launchIconPickerGridForSelectedPack(selectedIconPack);
            }
          }
        }).create();
  }

  void setDefaultIcon() {
    AppDetail app = apps.get(selectedAppPosition);
    IconPrefs iconPrefs = iconPrefsDao.getIconPrefs(app.getName());

    if (iconPrefs != null) {
      iconPrefsDao.deleteIconPrefs(iconPrefs);
    }

    try {
      Drawable icon = getPackageManager().getApplicationIcon(app.getName());
      changeIconInGrid(icon);
      changeIconInDialog(icon);

    } catch (PackageManager.NameNotFoundException e) {
      e.printStackTrace();
    }
  }

  void launchIconPickerGridForSelectedPack(String selectedIconPack) {
    Intent intent = new Intent(getBaseContext(), IconsActivity.class);
    intent.putExtra("iconpack", selectedIconPack);

    startActivityForResult(intent, 2);
  }

  @Override
  protected void onActivityResult(int requestCode, int resultCode, Intent data) {
    if (requestCode == 1) {
      loadApps();
      loadAppGrid();
      loadBackground();
    } else if (requestCode == 2) {
      if (resultCode == RESULT_OK) {
        String customIcon = data.getStringExtra("customicon");
        changeIcon(customIcon);
      }
    }
  }

  public void changeIcon(String customIcon) {
    String iconProp[] = customIcon.split("/");
    Drawable icon = new IconPackManager(this).loadDrawable(iconProp[0], iconProp[1]);

    changeIconInGrid(icon);
    changeIconInDialog(icon);
    changeIconInDatabase(customIcon);
  }

  void changeIconInGrid(Drawable icon) {
    apps.get(selectedAppPosition).setIcon(icon);
    adapter.notifyItemChanged(selectedAppPosition);
  }

  void changeIconInDialog(Drawable icon) {
    Bitmap bitmap = ((BitmapDrawable) icon).getBitmap();
    icon = new BitmapDrawable(getResources(), Bitmap.createScaledBitmap(bitmap, 200, 200, true));
    image.setImageDrawable(icon);
  }

  void changeIconInDatabase(String customIcon) {
    String appName = apps.get(selectedAppPosition).getName();
    IconPrefs iconPrefs = iconPrefsDao.getIconPrefs(appName);

    if (iconPrefs == null) {
      iconPrefs = new IconPrefs(appName);
    }

    iconPrefs.setIconName(customIcon);
    iconPrefsDao.addIconPrefs(iconPrefs);
  }

  public void changeLabel() {
    //TODO: Implement custom label on app grid
  }

  @Override
  public void onBackPressed() {
  }
}

File: app/src/main/java/com/djacoronel/sleeklauncher/home/MainActivity.kt
package com.djacoronel.sleeklauncher.home

import android.app.Activity
import android.app.ActivityOptions
import android.app.WallpaperManager
import android.content.*
import android.content.Intent.EXTRA_UID
import android.content.pm.PackageManager
import android.content.pm.ResolveInfo
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.support.v7.widget.GridLayoutManager
import android.util.DisplayMetrics
import android.view.View
import android.view.WindowManager
import android.widget.ImageView
import com.djacoronel.sleeklauncher.R
import com.djacoronel.sleeklauncher.data.model.AppDetail
import com.djacoronel.sleeklauncher.data.model.IconPrefs
import com.djacoronel.sleeklauncher.data.room.IconPrefsDao
import com.djacoronel.sleeklauncher.iconutils.IconPackManager
import com.djacoronel.sleeklauncher.iconutils.IconsActivity
import com.djacoronel.sleeklauncher.settings.BlurBuilder
import com.djacoronel.sleeklauncher.settings.SettingsActivity
import dagger.android.AndroidInjection
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.app_grid.*
import org.jetbrains.anko.alert
import org.jetbrains.anko.selector
import org.jetbrains.anko.toast
import java.util.*
import javax.inject.Inject

class MainActivity : Activity() {
    @Inject
    lateinit var iconPrefsDao: IconPrefsDao
    @Inject
    lateinit var preferences: SharedPreferences
    @Inject
    lateinit var iconPackManager: IconPackManager

    private lateinit var adapter: GridAdapter
    private lateinit var dialogIcon: ImageView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        AndroidInjection.inject(this)
        setContentView(R.layout.activity_main)

        loadBackground()
        setupBackgroundRefreshing()
        loadApps()
        setupGridRefreshing()

        // This makes status bar and navigation bar transparent
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            window.setFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
                    WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS)
        }
    }

    private fun loadBackground() {
        val bgPref = preferences.getString("bgPref", "0,0,0,0,0")
        val argbBlur = bgPref.split(",")
        val argb = intArrayOf(argbBlur[0].toInt(), argbBlur[1].toInt(), argbBlur[2].toInt(), argbBlur[3].toInt())
        val blur = argbBlur[4].toInt()

        loadBackgroundWithBlur(argb, blur)
    }

    private fun loadBackgroundWithBlur(argb: IntArray, blur: Int) {
        val wallpaperDrawable = WallpaperManager.getInstance(this).drawable

        var blurredBitmap =
                if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) (wallpaperDrawable as BitmapDrawable).bitmap
                else cropWallpaper((wallpaperDrawable as BitmapDrawable).bitmap)

        val numberOfFullBlur = blur / 25
        val remainingBlur = blur % 25

        for (i in 0 until numberOfFullBlur) blurredBitmap = BlurBuilder().blur(this, blurredBitmap, 25f)
        if (remainingBlur != 0) blurredBitmap = BlurBuilder().blur(this, blurredBitmap, remainingBlur.toFloat())

        mainBackground.visibility = View.VISIBLE
        mainBackground.setColorFilter(Color.argb(argb[0], argb[1], argb[2], argb[3]))
        mainBackground.setImageDrawable(BitmapDrawable(resources, blurredBitmap))
        mainBackground.scaleType = ImageView.ScaleType.CENTER_CROP
    }

    private fun cropWallpaper(wallpaper: Bitmap): Bitmap {
        val displayMetrics = DisplayMetrics()
        windowManager.defaultDisplay.getRealMetrics(displayMetrics)
        val x = 0
        val y = 0
        val height = displayMetrics.heightPixels
        val width = displayMetrics.widthPixels

        return if (wallpaper.width < (x + width) || wallpaper.height < (y + height))
            wallpaper
        else
            Bitmap.createBitmap(wallpaper, x, y, width, height)
    }

    private fun setupBackgroundRefreshing() {
        val intentFilter = IntentFilter()
        intentFilter.addAction(Intent.ACTION_WALLPAPER_CHANGED)

        val br = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                loadBackground()
            }
        }

        registerReceiver(br, intentFilter)
    }

    private fun loadApps() {
        val showHidden = preferences.getBoolean("showHidden", false)
        val selectedIconPack = preferences.getString("iconPack", "")

        val manager = packageManager
        val availableActivities = getInstalledAppsInfo()
        val apps = mutableListOf<AppDetail>()

        for (ri in availableActivities) {
            val name = ri.activityInfo.packageName
            val activityName = ri.activityInfo.name
            val label = ri.loadLabel(manager) as String
            val icon = ri.loadIcon(manager)

            val app = AppDetail(label, name, activityName, icon)
            val iconPrefs = iconPrefsDao.getIconPrefs(activityName)

            iconPrefs?.let {
                if (!it.isHidden || showHidden) {
                    if (it.iconName != IconPrefs.NO_CUSTOM_ICON) app.icon = iconPackManager.getAppIcon(it)
                    else app.icon = iconPackManager.getAppIcon(app, selectedIconPack)

                    if (it.label != IconPrefs.NO_CUSTOM_LABEL) app.label = it.label
                    apps.add(app)
                }
            }

            if (iconPrefs == null) {
                app.icon = iconPackManager.getAppIcon(app, selectedIconPack)
                apps.add(app)
            }
        }

        Collections.sort(apps) { a1, a2 -> a1.label.compareTo(a2.label, true) }
        loadAppGrid(apps)
    }

    private fun getInstalledAppsInfo(): List<ResolveInfo> {
        val intent = Intent(Intent.ACTION_MAIN, null)
        intent.addCategory(Intent.CATEGORY_LAUNCHER)
        return packageManager.queryIntentActivities(intent, 0)
    }

    private fun loadAppGrid(apps: List<AppDetail>) {
        adapter = GridAdapter(apps, this)
        app_grid.layoutManager = GridLayoutManager(this, 4)
        app_grid.adapter = adapter
    }

    private fun setupGridRefreshing() {
        val intentFilter = IntentFilter()
        intentFilter.addAction(Intent.ACTION_PACKAGE_REMOVED)
        intentFilter.addAction(Intent.ACTION_PACKAGE_ADDED)
        intentFilter.addDataScheme("package")
        val br = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                loadApps()
                val uid = intent.getIntExtra(EXTRA_UID, 0)
                val name = packageManager.getNameForUid(uid)
            }
        }
        this.registerReceiver(br, intentFilter)
    }

    fun launchApp(app: AppDetail, v: View) {
        val manager = packageManager
        val intent = manager.getLaunchIntentForPackage(app.name)
        val optsBundle = getMarshmallowOpeningAnimationOpts(v).toBundle()

        if (intent != null) {
            startActivity(intent, optsBundle)
        } else {
            toast("App not found! :(")
            adapter.removeApp(app.activity)
        }
    }

    private fun getMarshmallowOpeningAnimationOpts(v: View): ActivityOptions {
        val left = 0
        val top = 0
        val width = v.measuredWidth
        val height = v.measuredHeight

        // Use reveal animation if Marshmallow
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
            ActivityOptions.makeClipRevealAnimation(v, left, top, width, height)
        else
            ActivityOptions.makeScaleUpAnimation(v, left, top, width, height)
    }

    fun iconLongClick(app: AppDetail) {
        dialogIcon = ImageView(this)
        val bitmap = (app.icon as BitmapDrawable).bitmap
        val icon = BitmapDrawable(resources, Bitmap.createScaledBitmap(bitmap, 200, 200, true))
        dialogIcon.setImageDrawable(icon)
        dialogIcon.setOnClickListener { showIconPackList(app) }

        alert {
            title = "Options"
            message = "Tap icon to change it."
            isCancelable = true
            customView = dialogIcon
            negativeButton("Uninstall") {
                val packageUri = Uri.parse("package:" + app.name)
                val uninstallIntent = Intent(Intent.ACTION_DELETE, packageUri)
                startActivity(uninstallIntent)
            }

            val iconPrefs = iconPrefsDao.getIconPrefs(app.activity)
            if (iconPrefs == null)
                positiveButton("Hide") { hideIcon(IconPrefs(app.name, app.activity)) }
            else {
                if (!iconPrefs.isHidden) positiveButton("Hide") { hideIcon(iconPrefs) }
                else positiveButton("Show") { unHideIcon(iconPrefs) }
            }
        }.show()
    }

    private fun hideIcon(iconPrefs: IconPrefs) {
        iconPrefs.isHidden = true
        iconPrefsDao.addIconPrefs(iconPrefs)
        toast("App marked as hidden!")

        val showHidden = preferences.getBoolean("showHidden", false)
        if (!showHidden) adapter.removeApp(iconPrefs.activity)
    }

    private fun unHideIcon(iconPrefs: IconPrefs) {
        iconPrefs.isHidden = false
        iconPrefsDao.addIconPrefs(iconPrefs)
        toast("App removed from hidden!")
    }

    private fun showIconPackList(app: AppDetail) {
        val iconPacks = iconPackManager.availableIconPacks
        val iconPackNames = iconPacks.keys.toList()

        selector("Icon Packs", iconPackNames, { _, index ->
            if (iconPackNames[index] == "Default") {
                setDefaultIcon(app)
            } else {
                val selectedIconPack = iconPackNames[index]
                launchIconPickerGridForSelectedPack(app.activity, selectedIconPack)
            }
        })
    }

    private fun setDefaultIcon(app: AppDetail) {
        val iconPrefs = iconPrefsDao.getIconPrefs(app.activity)
        iconPrefs?.let { iconPrefsDao.deleteIconPrefs(it) }

        try {
            val icon = packageManager.getApplicationIcon(app.name)
            setIcon(app.activity, icon)
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }
    }

    private fun setIcon(activityName: String, icon: Drawable) {
        val app = adapter.getApp(activityName)

        app.icon = icon
        adapter.notifyDataSetChanged()

        val bitmap = (icon as BitmapDrawable).bitmap
        val dialogDrawable = BitmapDrawable(resources, Bitmap.createScaledBitmap(bitmap, 200, 200, true))
        dialogIcon.setImageDrawable(dialogDrawable)
    }

    private fun launchIconPickerGridForSelectedPack(activityName: String, selectedIconPack: String) {
        val intent = Intent(baseContext, IconsActivity::class.java)
        intent.putExtra("iconPack", selectedIconPack)
        intent.putExtra("activityName", activityName)
        startActivityForResult(intent, 2)
    }

    fun openSettings() {
        val intent = Intent(this, SettingsActivity::class.java)
        startActivityForResult(intent, 1)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == 1) {
            loadApps()
            loadBackground()
        } else if (requestCode == 2) {
            if (resultCode == Activity.RESULT_OK) {
                data?.let {
                    val customIcon = it.getStringExtra("customIcon")
                    val activityName = it.getStringExtra("activityName")
                    val iconProp = customIcon.split("/")
                    val icon = iconPackManager.loadDrawable(iconProp[0], iconProp[1])

                    val app = adapter.getApp(activityName)
                    val iconPrefs = IconPrefs(app.name, app.activity)
                    iconPrefs.iconName = customIcon
                    iconPrefsDao.addIconPrefs(iconPrefs)

                    setIcon(activityName, icon!!)
                }
            }
        }
    }

    override fun onBackPressed() {}
}
